src/main/java/uk/gov/nationalarchives/pdi/step/atomics/await/AwaitStep.java

```diff
@@ -148,89 +155,134 @@ public boolean processRow(final StepMetaInterface smi, final StepDataInterface s
+                    logLineNumber();
+                    return true; // row done!
                 }
-            } catch (final InterruptedException e) {
-                Thread.currentThread().interrupt(); // restore interrupted flag
-                // send row to the error output of the step
-                final String errorMessage = "Thread interrupted whilst waiting for Atomic value for id: " + atomicId;
-                this.putError(data.getOutputRowMeta(), row, 1L, errorMessage, data.getAtomicIdFieldName(), ErrorCodes.AWAIT_ATOMIC_WAIT_INTERRUPTED.getCode());
-                logLineNumber();
-                return true; // row done!
-            }
-        }  // end while
 
-        // send to specific target for Await success
-        final String metaAtomicValueTargetStepName = meta.getAtomicValueTargetStep() != null ? meta.getAtomicValueTargetStep().getName() : meta.getAtomicValueTargetStepname();
-        if (isNotEmpty(metaAtomicValueTargetStepName)) {
+                // refresh the atomic object
+                atomicValue = data.getAtomic(atomicId, atomicType);
+
+            }  // end while
+        }
+
+        if (awaitTarget != null) {
+            // send to specific target for Await success
+            final Set<RowSet> atomicValueTargetRowSets = data.getAtomicValueOutputRowSets().get(awaitTarget.getAtomicValue());
+            if (atomicValueTargetRowSets == null || atomicValueTargetRowSets.isEmpty()) {
+                throw new KettleException(BaseMessages.getString(PKG, "AwaitStep.Log.UnableToFindTargetRowSetForStep", new Object[] { awaitTarget.getTargetStep() != null ? awaitTarget.getTargetStep().getName() : awaitTarget.getTargetStepname() }));
+            }
 
-            this.putRowTo(data.getOutputRowMeta(), row, data.getAtomicValueOutputRowSet());
+            for (final RowSet awaitValueTargetRowSet : atomicValueTargetRowSets) {
+                this.putRowTo(data.getOutputRowMeta(), row, awaitValueTargetRowSet);
+            }
 
-            this.logDebug("Await DONE: <{0}>[{1}]", atomicId, meta.getAtomicValue());
-            logLineNumber();
+            this.logDebug("Await DONE: <{0}>[{1}]", atomicId, strNullIfNull(nullIfEmpty(awaitTarget.getAtomicValue())));
 
-            if (meta.isDiscardAtomic()) {
+            if (awaitTarget.isDiscardAtomic()) {
                 // discard the atomic if requested to do so
                 if (!data.removeAtomic(atomicId)) {
                     this.logError("Unable to discard Atomic with ID: {0}", atomicId);
                 }
             }
 
-            return true;  // row done!
+
         } else {
-            // raise an exception
-            throw new KettleException(BaseMessages.getString(PKG, "AwaitStep.Log.NoAtomicValueTargetStep"));
+            //send to default output if no Await Target
+            this.putRow(data.getOutputRowMeta(), row);
         }
+
+        logLineNumber();
+
+        return true;  // row done!
```

Comment:
Is it possible to abstract out any of this functionality? The `processRow` function is now over 200 lines long and this makes it hard to read and hard to test. Maybe we could put a ticket in the backlog to discuss and refactor this as without unit tests I think it is going to be difficult to debug and maintain in future.

Reply:
I agree it is a bit of a beast, my fear was that refactoring it was (a) difficult due to how the loops need to update the state of several function scoped variables, and (b) that refactoring it may actually make how it works less clear (I did try and include comments to help).

However, I am a believer that most code can be improved. So I would be quite happy to have some discussions and attempts to see if we could refactor this whilst keeping it as simple and clear. Perhaps we should involve @userjd in such discussions too.

Reply:
I think it can be refactored without too much difficulty. For example lines 221 - 250 could be replaced with something like `if(waitAndCheckAgain(meta,data,row,atomicId)) { return true; }` with those lines being abstracted to a `waitAndCheckAgain` function - this also removes the need for the comment :)

Reply:
I think `processRow` should now be in much better shape